### PR TITLE
Revise e-mail sending

### LINKED ISSUE #14: Revise how mail sending works
I think the API for `MailSender` is a bit strange. Sure, it works, but let's look at a "normal" use case for sending an e-mail at the moment:

1. Create model with information
2. Pass model and view name to the `IRazorEmailRenderer`
3. Pass the rendered mail, the sender and the recipient mail to a `MailSender` constructor
4. Call `SendAsync` on it

I think step 3 can be simplified (or rather generalized), because the sender doesn't really change between mails.

My thinking is that a different API might look like this:

```csharp
IRazorEmailRenderer renderer = ...;
SmtpAccount account = ...;

var sender = new MailKitMailSender(account);
sender.DefaultFrom = new MailAddress("no-reply@mycompany.com", "My Comany");
sender.DefaultReplyTo.Add(new MailAddress("support@mycompany.com", "My Company Support"));
sender.DefaultBcc.Add(new MailAddress("shadow-copy@mycompany.com"));

var model = new ExampleModel();
var body = await renderer.RenderAsync("MyMail", model);
var recipient = new MailAddress("you@example.com", "You");

var mail = new Mail
{
    Recipient = recipient,
    Body = body,
};

await sender.SendAsync(mail);
```

Things to note here are:
- the `MailSender` implementation gets an instance of the SMTP account instead of the mail details
- the sender has fields for default from, reply-to, bcc and cc (not shown here)
- a new `Mail` class that holds the rendered mail body and the recipient (not specifying from, cc, bcc or reply-to)
- the sender just has a method that accepts that mail object and sends it using the information given earlier

The details will need to be discussed, but I think this API is better as the sender can be easily reused and the `Mail` class would provide an extension point for future development.

### PR DESCRIPTION
**Pull request by ricardoboss**
Fixes #14

### CODE REVIEW

File: TRENZ.Lib.RazorMail.Core/Core/MailTemplateBase.cs [hunk 1]
```diff
@@ -38,52 +43,60 @@ public void AttachFile(string filename)
...
-        {
-            var attachment = new MailAttachment(fileData, filename, contentType);
-
-            switch (contentDisposition)
-            {
-                case ContentDisposition.Inline:
-                    attachment.ContentId = filename;
-                    attachment.Inline = true;
+        var cid = $"cid:{filename}";
+        if (Attachments.ContainsKey(filename))
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.Core/Core/MailTemplateBase.cs:71]: possibly useful, but also possibly not what the user intended; what if they want to attach two files with the same name? (and definitely scope creep)

File: TRENZ.Lib.RazorMail.Core/Core/MailTemplateBase.cs [hunk 2]
```diff
@@ -38,52 +43,60 @@ public void AttachFile(string filename)
...
-                    attachment.Inline = false;
-
-                    break;
-                default:
-                    break;
-            }
+        var attachment = new MailAttachment
+        {
+            FileName = filename,
+            FileStream = stream,
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.Core/Core/MailTemplateBase.cs:77]: how do you close the stream?

File: TRENZ.Lib.RazorMail.Core/Interfaces/IMailClient.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+using System.Threading.Tasks;
+
+using TRENZ.Lib.RazorMail.Models;
+
+namespace TRENZ.Lib.RazorMail.Interfaces;
+
+public interface IMailClient
+{
+    /// <summary>
+    /// The default "From" address.
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.Core/Interfaces/IMailClient.cs:12]: maybe add explanations what the Default* properties are useful for (and that you can override them per-message)

File: TRENZ.Lib.RazorMail.Core/Models/MailHeaderCollection.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    /// <summary>
+    /// The sender of the mail message.
+    /// </summary>
+    public MailAddress? From
+    {
+        get
+        {
+            if (ContainsKey("From"))
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.Core/Models/MailHeaderCollection.cs:55]: shouldn't these use `TryGetValue` to avoid double lookup?

File: TRENZ.Lib.RazorMail.Core/Models/MailHeaderCollection.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace TRENZ.Lib.RazorMail.Models;
+
+/// <summary>
+/// Represents a collection of mail headers.
+/// </summary>
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.Core/Models/MailHeaderCollection.cs:9]: this doesn't actually support custom headers yet, does it?

File: TRENZ.Lib.RazorMail.Core/Models/SmtpAccount.cs [hunk 6]
```diff
@@ -1,5 +1,37 @@
...
-                          string Login, string Password);
+/// <summary>
+/// Represents an SMTP account.
+/// </summary>
+public record SmtpAccount
+{
+    /// <summary>
+    /// The configuration section name.
+    /// </summary>
+    public const string SectionName = "SmtpAccount";
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.Core/Models/SmtpAccount.cs:11]: maybe a bit tightly coupled

File: TRENZ.Lib.RazorMail.Core/Services/BaseMailClient.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        if (message.Headers.From is null)
+            throw new InvalidOperationException("The 'From' address must be set.");
+
+        if (!message.Headers.Recipients.Any())
+            throw new InvalidOperationException("At least one recipient must be specified.");
+
+        if (string.IsNullOrWhiteSpace(message.Content.HtmlBody))
+            throw new InvalidOperationException("The HTML body must be set.");
+
+        if (string.IsNullOrWhiteSpace(message.Content.Subject))
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.Core/Services/BaseMailClient.cs:83]: I guess we could allow empty subjects?

File: TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs [hunk 8]
```diff
@@ -2,10 +2,10 @@
 
 using TRENZ.Lib.RazorMail.Models;
 
-namespace TRENZ.Lib.RazorMail.MailKitExtensions;
+namespace TRENZ.Lib.RazorMail.Extensions;
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs:5]: hmm, I had deliberately named it that because of overlapping / too-generic type names like MailboxAddress. This clarifies that those extensions are specifically for that particular implementation

File: TRENZ.Lib.RazorMail.SampleWebApi/Controllers/MailController.cs [hunk 9]
```diff
@@ -1,68 +1,62 @@
...
-        var mail = new SystemNetMailSender(
-            from: request.From,
-            to: request.To.Select(x => new MailAddress(x)),
-            cc: [],
-            bcc: [],
-            replyTo: [],
-            renderedMail: renderedMail,
-            logger: loggerFactory.CreateLogger<SystemNetMailSender>()
-        );
+        var message = new MailMessage
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.SampleWebApi/Controllers/MailController.cs:20]: can't all of this be abstracted now since the concrete impl comes from the param?

File: TRENZ.Lib.RazorMail.System.Net/Extensions/SmtpClientExtensions.cs [hunk 10]
```diff
@@ -1,40 +1,40 @@
...
-    }
+﻿using System;
+using System.Net.Mail;
+using System.Threading.Tasks;
+
+namespace TRENZ.Lib.RazorMail.Extensions;
+
+public static class SmtpClientExtensions
+{
+    public static Task SendAsync(this SmtpClient client, MailMessage message)
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.System.Net/Extensions/SmtpClientExtensions.cs:9]: add a `CancellationToken` with a TODO here, maybe

File: TRENZ.Lib.RazorMail.System.Net/SystemNetMailClient.cs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+using TRENZ.Lib.RazorMail.Services;
+
+using RazorMailMessage = TRENZ.Lib.RazorMail.Models.MailMessage;
+using SystemNetMailMessage = System.Net.Mail.MailMessage;
+
+namespace TRENZ.Lib.RazorMail;
+
+public class SystemNetMailClient(IOptions<SmtpAccount> accountOptions, ILogger<SystemNetMailClient> logger) : BaseMailClient(accountOptions)
+{
+
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.System.Net/SystemNetMailClient.cs:23]: extraneous linebreak

File: TRENZ.Lib.RazorMail.System.Net/SystemNetMailClient.cs [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        cancellationToken.ThrowIfCancellationRequested();
+
+        logger.LogInformation("Sending mail from {From} to {Recipients} (CC: {Cc}, BCC: {Bcc}) with subject {Subject}",
+            nativeMessage.From, nativeMessage.To, nativeMessage.CC, nativeMessage.Bcc, nativeMessage.Subject);
+
+        await client.SendAsync(nativeMessage);
+    }
+
+    [MustDisposeResource]
+    private static SystemNetMailMessage ConvertToNative(RazorMailMessage razorMessage)
```
Comment (chucker) [inline TRENZ.Lib.RazorMail.System.Net/SystemNetMailClient.cs:51]: maybe instead make this an extension method as well (`razorMessage.ToMailMessage()`)?

**Review by chucker — COMMENTED** [review]
overall looks good. Some scope creep; maybe make the switch to `Stream` a separate PR, for example.

File: TRENZ.Lib.RazorMail.Core/Core/MailTemplateBase.cs (hunk 1 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.Core/Core/MailTemplateBase.cs:71]: I didn't even know this was possible with mail attachments. I think I didn't change the logic, so this wasn't possible before either. The only thing I changed was the nesting to improve readability.

File: TRENZ.Lib.RazorMail.Core/Core/MailTemplateBase.cs (hunk 2 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.Core/Core/MailTemplateBase.cs:77]: Yeah, I'll revert to using byte[] for now

File: TRENZ.Lib.RazorMail.Core/Interfaces/IMailClient.cs (hunk 3 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.Core/Interfaces/IMailClient.cs]: Done

File: TRENZ.Lib.RazorMail.Core/Models/MailHeaderCollection.cs (hunk 4 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.Core/Models/MailHeaderCollection.cs]: Yeah, fixed

File: TRENZ.Lib.RazorMail.Core/Models/MailHeaderCollection.cs (hunk 5 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.Core/Models/MailHeaderCollection.cs:9]: Yes it does, it is just a dictionary after all. The mail client implementation just don't use other keys yet. Should be add this in this PR too?

File: TRENZ.Lib.RazorMail.Core/Models/SmtpAccount.cs (hunk 6 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.Core/Models/SmtpAccount.cs:11]: It's part of the Options pattern. What do you propose?

File: TRENZ.Lib.RazorMail.Core/Services/BaseMailClient.cs (hunk 7 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.Core/Services/BaseMailClient.cs]: Sure

File: TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs (hunk 8 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs:5]: Any reason why the mailkit and systemnet implementations don't have their own namespaces?

File: TRENZ.Lib.RazorMail.SampleWebApi/Controllers/MailController.cs (hunk 9 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.SampleWebApi/Controllers/MailController.cs:20]: Yes

File: TRENZ.Lib.RazorMail.System.Net/Extensions/SmtpClientExtensions.cs (hunk 10 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.System.Net/Extensions/SmtpClientExtensions.cs]: I implemented it right away. The token calls `SendAsyncCancel` on the client and if I understand it correctly, the TCS will be marked as cancelled by the `SendCompleted` handler.

File: TRENZ.Lib.RazorMail.System.Net/SystemNetMailClient.cs (hunk 11 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.System.Net/SystemNetMailClient.cs]: Fixed

File: TRENZ.Lib.RazorMail.System.Net/SystemNetMailClient.cs (hunk 12 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.System.Net/SystemNetMailClient.cs]: I moved all conversion methods to extension classes

File: TRENZ.Lib.RazorMail.Core/Core/MailTemplateBase.cs (hunk 1 above)
Comment (chucker) [reply TRENZ.Lib.RazorMail.Core/Core/MailTemplateBase.cs]: My bad.

This behavior is probably wrong; see #21.

File: TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs (hunk 8 above)
Comment (chucker) [reply TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs:5]: Is your question "…why the… implementations _should_ have their own namespaces"?

File: TRENZ.Lib.RazorMail.Core/Models/SmtpAccount.cs (hunk 6 above)
Comment (chucker) [reply TRENZ.Lib.RazorMail.Core/Models/SmtpAccount.cs:11]: I guess that's fine.

Maybe move it to `Models/Configuration/` to make that clearer?

(Some of the folder structure isn't very intuitive, unrelated to this PR.)

File: TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs (hunk 8 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs:5]: No, the opposite. Why didn't they have their own namespaces already? Any reason? Otherwise I'd just add them

File: TRENZ.Lib.RazorMail.Core/Models/SmtpAccount.cs (hunk 6 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.Core/Models/SmtpAccount.cs:11]: Sure

File: TRENZ.Lib.RazorMail.Core/Models/MailHeaderCollection.cs (hunk 5 above)
Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.Core/Models/MailHeaderCollection.cs:9]: Kinda? The keys aren't hardcoded in the implementations though, they use the getters in the collection which in turn have "hardcoded"/known keys.

File: TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs (hunk 8 above)
Comment (chucker) [reply TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs:5]: If the `BaseSmtpMailClient` implementations come with their own namespace, it would be `TRENZ.Lib.RazorMail.MailKit.MailKitMailClient`, which seems a bit verbose to me?

Comment (ricardoboss) [reply TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs]: Yeah, but you virtually never write out the full name. IMHO it's better to be verbose than to be imprecise.

Comment (chucker) [reply TRENZ.Lib.RazorMail.MailKit/Extensions/MailAddressExtensions.cs]: > it's better to be verbose than to be imprecise

agreed

**Review by chucker — APPROVED** [review]
